File: lib/commands/xkcd.rb
# frozen_string_literal: true

module Rebbot
  module Commands
    class XKCD < Rebbot::Commands::Base
      on :xkcd, description: 'xkcd comic'

      def with_options(cmd)
        cmd.integer('number', 'comic number (omit for random)')
      end

      def on_event(event)
        number = event.options['number'] || rand(1..latest['num'])
        json = comic(number)
        event.respond(embeds: [
          {
            title: "#{number}: #{json['safe_title']}",
            url: "https://xkcd.com/#{number}",
            image: { url: json['img'] },
            description: json['alt']
          }
        ])
      rescue Faraday::Error, JSON::ParserError => e
        if e.is_a?(Faraday::ResourceNotFound)
          event.respond(content: '🤷 Comic not found')
        else
          event.respond(content: "💀 An error occurred: #{e}, help! <@#{Rebbot::ROB_ID}>")
        end
      end

      private

      def comic(number = nil)
        response = xkcd.get("/#{number}/info.0.json")
        JSON.parse(response.body)
      end

      def latest
        response = xkcd.get('/info.0.json')
        JSON.parse(response.body)
      end

      def xkcd
        Faraday.new(
          url: 'https://xkcd.com',
          headers: { 'Content-Type' => 'application/json' }
        ) do |c|
          c.use Faraday::Response::RaiseError
        end
      end
    end
  end
end

File: src/commands/xkcd.ts
import { InteractionResponseType } from "discord-interactions";
import { CommandOptionType, cmd } from "../types.js";

export default cmd(
  {
    name: "xkcd",
    description: "xkcd comic retriever",
    options: [
      {
        name: "number",
        description: "Comic number (random if omitted)",
        type: CommandOptionType.INTEGER,
        required: false,
        min_value: 1,
      },
    ],
  },
  async ({ number: num }) => {
    try {
      if (num == null) {
        const latest = (await fetch("https://xkcd.com/info.0.json").then((r) =>
          r.json(),
        )) as { num: number };
        num = Math.floor(Math.random() * latest.num) + 1;
      }

      const comic = (await fetch(`https://xkcd.com/${num}/info.0.json`).then(
        (r) => {
          if (!r.ok) throw new Error(`Comic ${num} not found`);
          return r.json();
        },
      )) as { title: string; num: number; img: string; alt: string };

      return {
        type: InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
        data: {
          embeds: [
            {
              title: `#${comic.num}: ${comic.title}`,
              url: `https://xkcd.com/${comic.num}`,
              image: { url: comic.img },
              footer: { text: comic.alt },
            },
          ],
        },
      };
    } catch {
      return {
        type: InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
        data: { content: "Failed to fetch xkcd comic." },
      };
    }
  },
);
